### PR TITLE
Reworked SSB

### PR DESCRIPTION
**Pull request by uellue**
* Passes all tests, equivalent to reference SSB with exception of numerical errors
* Faster than reference SSB and less memory consumption
* Use high-level numpy/cupy instructions to help with GPU porting
* Generally shortened and simplified
* Handle ROIs correctly
* Decouple reconstruction shape from dataset shape (to be validated for correctness)
* GPU(CuPy) support based on https://github.com/LiberTEM/LiberTEM/pull/760
* Adaptations for discrepancies between NumPy and CuPy
* Allow using an external mask container for using with InlineJobExecutor and single core live processing

For best results use an InlineJobExecutor and external MaskContainer since the mask stack is too large. :-)

## Contributor Checklist:

* [x] I have added myself to [the creators.json file](https://github.com/Ptychography-4-0/ptychography/blob/master/packaging/creators.json)
* [x] I have added [a changelog entry](https://github.com/Ptychography-4-0/ptychography/tree/master/docs/source/changelog) for my contribution
* [ ] I have added/updated documentation for all user-facing changes
* [x] I have added/updated test cases

### CODE REVIEW

**Comment by uellue** [comment]
It turns out that the mask stack has to be calculated very carefully and faithfully, with anti-aliasing and possibly taking all kinds of distortions into account to map detector coordinates and scattering angles faithfully.

This should be taken into account when designing an "on the fly" calculation.

File: src/ptychography/reconstruction/ssb.py [hunk 1]
```diff
@@ -84,19 +85,23 @@ def generate_masks(reconstruct_shape, mask_shape, dtype, wavelength, dpix, semic
...
+        antialiased=True
+    )
 
     half_reconstruct = (reconstruct_shape[0]//2 + 1, reconstruct_shape[1])
     masks = []
 
-    cos_angle = np.cos(np.pi / 180 * angle)
-    sin_angle = np.sin(np.pi / 180 * angle)
+    cos_angle = 1 # np.cos(np.pi / 180 * angle)
+    sin_angle = 0 # np.sin(np.pi / 180 * angle)
```
Comment (sk1p) [inline src/ptychography/reconstruction/ssb.py:99]: Temporarily disabled? Needs to be enabled again, right?

File: src/ptychography/reconstruction/ssb.py [hunk 2]
```diff
@@ -73,7 +74,7 @@ def rotate_sysx(real_sy, real_sx, cos_angle, sin_angle):
 
 
 def generate_masks(reconstruct_shape, mask_shape, dtype, wavelength, dpix, semiconv,
```
Comment (sk1p) [inline src/ptychography/reconstruction/ssb.py:76]: What is the unit of `cutoff`? I think `generate_masks` could use a docstring, also update of the UDF docstring for `cutoff` parameter.

Comment (sk1p) [reply src/ptychography/reconstruction/ssb.py:12]: Something else: the `wavelength` parameter shadows the imported `wavelength` function; one of these should be renamed

**Comment by uellue** [comment]
Testing implementations that are only approximately close to the reference SSB requires realistic test data. If the data is not realistic, the signal for each spatial frequency can cancel out to zero or build up unrealistically from summing many complex numbers with different phase angles. This can lead to seemingly very different results from only slightly different SSB implementations.

For that reason we need a method to generate realistic synthetic ptychography data.

@bangunarya recently you showed a neat notebook to test various ptychography methods. That included a simple method to generate input data. Would you perhaps like to contribute this code for generating data to the repository?

**Comment by uellue** [comment]
@bangunarya this PR is updated to use your test data now. The notebook is included as well in prototypes/ssb/ssb-validation.ipynb 

Could you perhaps simulate a dataset that is similar to the current MoS2 one, but with lighter atoms (carbon or even hydrogen) and with a larger semiconvergence angle? 100 mrad should work.

**Comment by bangunarya** [comment]
Hi @uellue, sure! Let me try to generate the ground truth.

**Comment by uellue** [comment]
The code now achieves a good match with simulated reference data and should have decent test coverage.

To facilitate further development, perhaps we can first merge this, then get CI and homepage working in this repo, and after that work on documentation and examples?

File: tests/reconstruction/test_ssb.py [hunk 3]
```diff
@@ -1,228 +1,625 @@
...
+
+try:
+    from libertem.utils.devices import detect, has_cupy
+    from libertem.common.backend import set_use_cpu, set_use_cuda
+    use_cupy = detect()['cudas'] and has_cupy()
+except ImportError:
+    use_cupy = False
+
+DATA_PATH = "/storage/holo/clausen/testdata/ER-C-1/projects/ptycho-4.0/data/RefData/slice_00002_thick_0.312293_nm_blocksz.mat"
+PARAM_PATH = "/storage/holo/clausen/testdata/ER-C-1/projects/ptycho-4.0/data/slice_00002_thick_0.312293_nm_blocksz.params.json"
```
Comment (SimeonEhrig) [inline tests/reconstruction/test_ssb.py:26]: @uellue 
Is there a way, that you can provide the test data? I want to test the implementation.

Comment (uellue) [reply tests/reconstruction/test_ssb.py:26]: @SimeonEhrig no problem! Does this link work? https://iffcloud.fz-juelich.de/s/ee72R8D9Rkbfrez

Comment (uellue) [reply tests/reconstruction/test_ssb.py:26]: We are preparing a better way to share this, perhaps Zenodo upload

**Comment by uellue** [comment]
Merged since PR #29 would benefit from the updates to the testing infrastructure and we can continute working on SSB with follow-up PRs.

**Pull request merged**